pkg/controller/grouppermission/grouppermission_controller.go

```diff
@@ -171,9 +178,139 @@ func (r *ReconcileGroupPermission) Reconcile(request reconcile.Request) (reconci
 		return reconcile.Result{}, nil
 	}
 
+	// ___________________Permission and Namespace logic_______________________ //
```

Comment:
With this change, please move some of the functionality into functions (or methods) as appropriate. `Reconcile` is now just shy of 200 lines, and it's hard to follow. A side effect will make it easier to test certain pieces in isolation.